src/commonMain/kotlin/fr/acinq/lightning/io/Peer.kt

```diff
@@ -272,24 +271,30 @@ class Peer(
     }
 
     fun connect() {
-        if (connectionState.value is Connection.CLOSED) establishConnection()
-        else logger.warning { "Peer is already connecting / connected" }
+        if (connectionState.value is Connection.CLOSED) {
+            _connectionState.value = Connection.ESTABLISHING
+            establishConnection()
+        } else {
+            logger.warning { "Peer is already connecting / connected" }
+        }
     }
 
     fun disconnect() {
         if (this::socket.isInitialized) socket.close()
         _connectionState.value = Connection.CLOSED(null)
-        // TODO in exception handler? output.close()
     }
 
     // Warning : lateinit vars have to be used AFTER their init to avoid any crashes
     //
-    // This shouldn't be used outside the establishedConnection() function
+    // This shouldn't be used outside the establishConnection() function
     // Except from the disconnect() one that check if the lateinit var has been initialized
     private lateinit var socket: TcpSocket
     private fun establishConnection() = launch {
+        // Clean up previous connection state: we do this here to ensure that it is handled before the Connected event for the new connection.
```

Comment:
Don't you feel like replicating the `ListenJob` pattern of `ElectrumClient`? And the `CoroutineExceptionHandler` snippet? If only for symmetry? Why would it be appropriate for electrum but not for the peer?